Add support for ENS contenthash methods

cgewecke opened the pull request:
## Description

Adds `getContenthash` and `setContenthash` methods to the ENS module.

**Example**
```js
await web3.eth.ens.getContenthash('ethereum.eth');

> {
  "protocolType": "ipfs",
  "decoded": "QmaEBknbGT4bTQiQoe2VNgBJbRfygQGktnaW5TbuKixjYL"
}

await web3.eth.ens.setContenthash(
  'ethereum.eth', 
  'ipfs://QmaEBknbGT4bTQiQoe2VNgBJbRfygQGktnaW5TbuKixjYL,
);
```
As discussed in #3392, there are now two widely used resolver contracts for ENS with different interfaces. 

~~PR adds support for the newer PublicResolver ABI, toggling between the two candidates by making an EIP 165 `supportsInterface` call for the `contenthash` method before selecting which ABI to use. If the Resolver address associated with the ENS name supports contenthash, the PublicResolver ABI is used, otherwise module defaults to the legacy Resolver ABI.~~

PR updates the ENS Resolver ABI to include contenthash interface as well as deprecated `content`. Makes EIP 165 supportsInterface calls for a subset of resolver methods to verify that the target contract implements them.

ENS has [a feature][1] that allows nodes to register ABIs on-chain for contract addresses but this does not appear to be implemented for the publicly deployed Resolver contracts themselves. 

Running a [reverse address check][2] on Etherscan for the mainnet PublicResolver at: [0x4976fb03C32e5B8cfe2b6cCB31c09Ba78EBaBa41][3] does not discover an ENS node which might supply an ABI. 


[1]: https://eips.ethereum.org/EIPS/eip-205
[2]: https://etherscan.io/enslookup
[3]: https://etherscan.io/address/0x4976fb03C32e5B8cfe2b6cCB31c09Ba78EBaBa41

Fixes #3392

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [x] I ran ```npm run dtslint``` with success and extended the tests and types if necessary.
- [x] I ran ```npm run test:unit``` with success.
- [ ] I have executed ``npm run test:cov`` and my test cases do cover all lines and branches of the added code.
- [ ] I ran ```npm run build-all``` and tested the resulting file/'s from  ```dist``` folder in a browser.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the live network.

cgewecke committed:
Upgrade ensdomains/resolver (dev dep) to 0.2.4

cgewecke committed:
Add PublicResolver ABI

cgewecke committed:
Select resolver ABI via EIP165 (Supports Interface)

cgewecke committed:
Update ens.js injection tests with additional resolver call

cgewecke committed:
Add error handling for mismatched resolver ABI & method calls

cgewecke committed:
Draft: get/set content hash

cgewecke committed:
Fix jshint warnings (where possible)

cgewecke committed:
Add "unsupported abi" injection tests for contenthash & setContenthash

cgewecke committed:
Update contentHash typescript & tests

cgewecke committed:
Update docs for getContentHash / setContentHash

cgewecke commented on packages/web3-eth-ens/src/lib/contentHash.js:
```diff
@@ -0,0 +1,85 @@
+/*
+  Adapted from ensdomains/ui
+  Permalink: https://github.com/ensdomains/ui/blob/3e62e440b53466eeec9dd1c63d73924eefbd88c1/src/utils/contents.js#L1-L85
+*/
```
Have opened an issue at [@ensdomains/ui 62][1] asking for clarification about how to license & attribute this file's contents appropriately. The repo currently has no license. 

[1]: https://github.com/ensdomains/ui/issues/62

ricmoo commented:
ABI is supported in the public resolver (resolver.eth) See line 72: https://etherscan.io/address/0x4976fb03c32e5b8cfe2b6ccb31c09ba78ebaba41#code

cgewecke commented:
@ricmoo I might misunderstand this but isn't that ABI method similar to `contentHash`, e.g it lets you associate an ABI with a name? 

In this case we're looking for the ABI *of the resolver itself* and I wasn't able to see a way of retrieving it. 

If it is possible it would be the best solution by far though...

ricmoo commented:
Oh! I see what you mean.

We could talk to @Arachnid about adding the ABI to resolver.eth, but I think it would be fine to just hardcode the getters (and maybe the setters) into it; each has an associated EIP.

Also, maybe it should be `contenthash`, not `contentHash` (and `setContenthash`, not `setContentHash`) to keep it inline with the [EIP-1577 standard](https://eips.ethereum.org/EIPS/eip-1577)?

cgewecke commented:
> Also, maybe it should be contenthash, not contentHash (and setContenthash, not setContentHash) to keep it inline with the standard?

@ricmoo Ah yes, that's a good point.

cgewecke committed:
Use single, extended Resolver ABI

cgewecke committed:
Call supportsInterface for each method

cgewecke committed:
Undo contenthash camel-casing (per EIP convention)

cgewecke committed:
Add ENS license to decoding utils

cgewecke committed:
Add E2E test for pubkey methods

cgewecke committed:
Fix Registry.js diff

cgewecke commented:
Have revised per initial review - think this is moving in a better direction...
 
+ Removed ABI toggling
+ Added logic to query supportsInterface for each resolver method call Web3 wraps and ENS [advertises an interface ID for in their docs][2]. This is tolerant - e.g it will error if the Resolver doesn't support `content` but ignore methods it doesn't know about.
+ Replaced ABI with the `Resolver` [generic][1] from @ensdomain/resolver@0.2.4. (This includes definitions for deprecated methods like `content`).
+ Added an E2E test for pubkey set/get - its interface is also getting validated now.
+ Removed camel-casing of `contenthash` 

[1]: https://github.com/ensdomains/resolvers/blob/master/contracts/Resolver.sol#L39-L43
[2]: https://docs.ens.domains/contract-developer-guide/writing-a-resolver

cgewecke committed:
Update method signatures in docs

nivida commented on package.json:
```diff
@@ -137,15 +139,13 @@
         "typescript": "next",
         "underscore": "^1.9.1",
         "vinyl-source-stream": "^2.0.0",
-        "wait-port": "^0.2.6",
```
there are unnecessary changes in this file as the wait-port and truffle dependency movement :-)

cgewecke replied:
The package.json has been normalized alphabetically in some other PR now. This isn't in the diff anymore.

nivida commented on packages/web3-eth-ens/package.json:
```diff
@@ -13,6 +13,7 @@
     },
     "main": "src/index.js",
     "dependencies": {
+        "content-hash": "^2.5.2",
```
I would recommend asking @ricmoo here again.. there are some cases this package wasn't doing his job nicely.. and also here.. is this external dependency required from a security perspective?

cgewecke replied:
This package is [used by @ensdomains/ui][1] for the same purpose. Intention is to make the decoding here as close to theirs as possible without pulling in lots of other deps. 

[1]: https://github.com/ensdomains/ui/blob/d2bfbe92fce7fc18e3caeb6ccec969de804b3f86/package.json#L60

nivida commented on packages/web3-eth-ens/src/lib/ResolverMethodHandler.js:
```diff
@@ -22,8 +22,22 @@
 
+// These ids obtained at ensdomains docs:
+// https://docs.ens.domains/contract-developer-guide/writing-a-resolver
+var interfaceIds = {
```
This should get out-sourced to the config file I think..

cgewecke replied:
Fixed in 
a6e19cb

nivida commented on packages/web3-eth-ens/src/lib/ResolverMethodHandler.js:
```diff
@@ -41,14 +55,15 @@ function ResolverMethodHandler(registry) {
  * @returns {Object}
  */
-ResolverMethodHandler.prototype.method = function (ensName, methodName, methodArguments, callback) {
+ResolverMethodHandler.prototype.method = function (ensName, methodName, methodArguments, outputFormatter, callback) {
```
The outputFormatter addition wouldn't be required.. you can just wrap the promise as on other places :-)

nivida replied:
(This ugly handler has to get refactored/removed anyways.. just ping me.. I do have a solution somewhere on a handwritten paper)

cgewecke replied:
Not sure I follow, but yes this could be refactored at some point.

nivida commented on packages/web3-eth-ens/src/lib/ResolverMethodHandler.js:
```diff
@@ -69,9 +84,11 @@ ResolverMethodHandler.prototype.call = function (callback) {
-    this.parent.registry.getResolver(this.ensName).then(function (resolver) {
-        self.parent.handleCall(promiEvent, resolver.methods[self.methodName], preparedArguments, callback);
+    this.parent.registry.getResolver(this.ensName).then(async function (resolver) {
+        await self.parent.checkInterfaceSupport(resolver, self.methodName);
```
I remember a discussion I had once with @cgewecke about additional network requests and their impact. Shouldn't this ``checkInterface`` call not also be optional with a ``strictEns`` flag or so? Otherwise, the arguments in the discussion we had are false:)

cgewecke replied:
The origin of this design is the discussion beginning in [#3392 comment][1]. In this case there are two widely deployed resolvers with different APIs, so there's a strong case for validating the request before-hand. 

I think you're alluding here to a [debate in #3428][1] about whether or not to gate revert-with-reason behind an option flag. The ENS module is not subject to the similar considerations AFAIK, personally don't think this will impact people in the same way 🤷  

[1]: https://github.com/ethereum/web3.js/pull/3248#pullrequestreview-324469285




[1]: https://github.com/ethereum/web3.js/issues/3392#issuecomment-591918724

nivida commented on packages/web3-eth-ens/src/lib/ResolverMethodHandler.js:
```diff
@@ -208,4 +230,31 @@ ResolverMethodHandler.prototype.prepareArguments = function (name, methodArgumen
+ * @returns {Promise}
+ */
+ResolverMethodHandler.prototype.checkInterfaceSupport = async function (resolver, methodName) {
+    // Skip validation for undocumented interface ids (ex: multihash)
```
FYI: Multihash is here for backward compatibility

nivida commented on packages/web3-eth-ens/src/lib/ResolverMethodHandler.js:
```diff
@@ -208,4 +230,31 @@ ResolverMethodHandler.prototype.prepareArguments = function (name, methodArgumen
+            .supportsInterface(interfaceIds[methodName])
+            .call();
+    } catch(err) {
+        console.warn('Could not verify interface of resolver contract at "' + resolver.options.address + '". ');
```
Couldn't it also throw other errors? ;-)

cgewecke replied:
Yes, it could but is spliced before another call to the resolver. If there's a problem with the connection or something, the error will be thrown then. I don't think this will swallow anything unexpectedly.

nivida commented on packages/web3-eth-ens/src/lib/contentHash.js:
```diff
@@ -0,0 +1,111 @@
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
```
Instead of just copy it should we probably create an EF package for it with working together with ENS @Arachnid

cgewecke replied:
This code comes out of [@ensdomains/ui][1] which is relatively large. This seemed like the lightest approach.

[1]: https://github.com/ensdomains/ui

nivida commented on packages/web3-eth-ens/src/ressources/ABI/Resolver.js:
```diff
@@ -1,23 +1,158 @@
 "use strict";
 
+// This ABI taken from @ensdomains/resolver@0.2.4
+// @ensdomains/resolver/build/contracts/Resolver.json
```
I once thought about using those ABI's from the ENS packages itself.. we could/should do that and add the additional required ABI items for backward compatibility with a clear deprecation guide/timeline. :-)

cgewecke replied:
Agree, that's better. Fixed in efd5cbf

nivida commented on packages/web3-eth-ens/src/ressources/ABI/Resolver.js:
```diff
@@ -34,33 +169,39 @@ var RESOLVER = [
         "outputs": [
             {
-                "name": "contentType",
+                "name": "",
```
Are those return values also unnamed in legacy Dapps?:)

cgewecke replied:
Fixed in efd5cbf

nivida commented on packages/web3-eth-ens/src/ressources/ABI/Resolver.js:
```diff
@@ -242,114 +495,153 @@ var RESOLVER = [
+        "payable": false,
+        "stateMutability": "nonpayable",
+        "type": "function"
     }
```
Overall do I think are there some unrelated changes in this ABI.. we should verify every change or undo the non-required changes.

cgewecke replied:
Fixed in efd5cbf

nivida commented on packages/web3-eth-ens/types/index.d.ts:
```diff
@@ -22,6 +22,12 @@ import { TransactionRevertInstructionError } from 'web3-core-helpers';
 import { Eth } from 'web3-eth';
 import { Contract } from 'web3-eth-contract';
 
+export interface ContentHash {
```
Noice! 🎊

nivida commented on test/e2e.ens.js:
```diff
@@ -50,5 +57,126 @@ describe('ENS [ @E2E ]', function () {
         );
     });
+
+    it('should get/set a publickey', async function(){
```
Some great new test cases thanks Chris! But also here.. aren't there unnecessary changes that would be against the guidelines of this project? :-)

cgewecke replied:
Idk. This PR introduces to interface validation as part of the solution to supporting contentHash per design discussion in #3392. We have existing methods that will be validated this way so we should also test them.

nivida commented on test/eth.ens.js:
```diff
@@ -103,6 +103,8 @@ function isExpectedReceipt(receipt) {
 }
 
 describe('ens', function () {
+    this.timeout(5000);
```
Can you elaborate on why exactly this overall timeout is required? 5s are quite a time :/

cgewecke replied:
Thanks for catching. Fixed in f037c31

cgewecke commented:
Rebased on bcf248f895fe14d466e6fe60e9096f0fc5e7a766

cgewecke committed:
Fix web3-eth-ens package-lock.json rebase break

cgewecke committed:
Remove extraneous mocha timeout in test/eth.ens.js

cgewecke committed:
Move interfaceIds to config

cgewecke committed:
Minimize changes to Resolver ABI

cgewecke commented:
Update: Have made suggested improvements per @nivida's review:
+ interfaceIds moved to config
+ preserved the legacy ABI and only added contentHash methods as necessary
+ deleted a stray mocha timeout

andrewheadricke commented:
Just stumbled across this issue, looking forward to getting this incorporated into web3

ryanio committed:
Merge branch '1.x' into issue/3392-ens-contenthash

ryanio reviewed (approved):
LGTM! Nice comprehensive work @cgewecke.

New tests are passing successfully in CI:
```
    ✓ should error when calling "getContent" if resolver does not support it
    ✓ should error when calling "setContent" if resolver does not support it
    ✓ getContenthash return object keys are null if no contentHash is set
    ✓ should get/set an IPFS contenthash (ipfs://)
    ✓ should get/set an IPFS contenthash (/ipfs/)
    ✓ should get/set a bzz contenthash
    ✓ should get/set an onion contenthash
    ✓ should get/set an onion3 contenthash
    ✓ setContenthash errors when encoding an invalid contenthash (promise)
    ✓ setContentHash errors when encoding an invalid contenthash (callback)
```
